Framework/NexusGeometry/test/NexusGeometrySaveTest.h

```diff
@@ -0,0 +1,1539 @@
+* test rotation of source written to file is same as in component info.
+
+LOCATION TESTS:
+* test NXmonitor location not written when is at origin
+* test NXdetector location not written when is at origin
+* test NXsource location not written when is at origin
+
+ROTATION TESTS
+* test NXdetector rotation not written when is zero
+* test NXmonitor rotation not written when is zero
+* test NXsource rotation not written when is zero
+
+====================================================================
+*/
+
+  void
+  test_when_nx_detector_groups_have_nx_transformations_transformation_type_is_specified_for_all() {
+
+    ScopedFileHandle fileResource(
+        "check_nxdetector_groups_have_transformation_types_test_file.hdf5");
+    std::string destinationFile = fileResource.fullPath();
+
+    auto instrument =
+        ComponentCreationHelper::createTestInstrumentRectangular2(2, 2);
+    auto instr = Mantid::Geometry::InstrumentVisitor::makeWrappers(*instrument);
+
+    // saveinstrument
+    NexusGeometrySave::saveInstrument(instr, destinationFile,
+                                      DEFAULT_ROOT_PATH);
+    auto &compInfo = (*instr.first);
+
+    auto instrName = compInfo.name(compInfo.root());
+
+    HDF5FileTestUtility tester(destinationFile);
+
+    for (size_t i = compInfo.root() - 1; i > 0; --i) {
+
+      if (Mantid::Geometry::ComponentInfoBankHelpers::isAnyBank(compInfo, i)) {
+
+        fullH5Path path = {DEFAULT_ROOT_PATH, instrName, compInfo.name(i),
+                           TRANSFORMATIONS};
+        bool hasNXTransformation =
+            tester.hasAttributeInGroup(NX_CLASS, NX_TRANSFORMATIONS, path);
+
+        // TODO: having such a group may be optional.
+        TS_ASSERT(hasNXTransformation);
+
+        bool hasTranslation = tester.hasDataset(LOCATION, path);
+
+        bool hasRotation = tester.hasDataset(ORIENTATION, path);
+
+        TS_ASSERT((hasRotation || hasTranslation));
+      }
+    }
+  }
+
+  void
+  test_when_nx_monitor_groups_have_nx_transformations_transformation_type_is_specified_for_all() {
+
+    ScopedFileHandle fileResource(
+        "check_nxmonitor_groups_have_transformation_types_test_file.hdf5");
+    std::string destinationFile = fileResource.fullPath();
+
+    const Quat someRotation(45.0, V3D(0.0, 1.0, 0.0));
+
+    auto instrument =
+        ComponentCreationHelper::createMinimalInstrumentWithMonitor(
+            V3D(0, 0, 0), someRotation);
+    auto instr = Mantid::Geometry::InstrumentVisitor::makeWrappers(*instrument);
+
+    // saveinstrument
+    NexusGeometrySave::saveInstrument(instr, destinationFile,
+                                      DEFAULT_ROOT_PATH);
+    auto &compInfo = (*instr.first);
+    auto &detInfo = (*instr.second);
+
+    auto instrName = compInfo.name(compInfo.root());
+
+    HDF5FileTestUtility tester(destinationFile);
+
+    auto detIds = detInfo.detectorIDs();
+
+    for (const int &ID : detIds) {
```

Comment:
loop looks to be routing through map for no reason. Why can't we just work with indexes directly....
`for(size_t i = 0: i < detInfo.size(); ++i)` ...

Reply:
refactored